[FIX] mass_mailing: convert base64 background images to inline images

Zinston opened the pull request:
Base64 img src were converted to inline images but background images were omitted. This commit fixes that.

opw-3374767
task-3425134

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/126794).

Zinston commented:
@tde-banana-odoo Would you mind reviewing this? It concerns this ticket: https://www.odoo.com/web#id=3374767&cids=1&menu_id=4720&action=333&active_id=1695&model=project.task&view_type=form Thanks!

Zinston commented on addons/mass_mailing/models/mailing.py:
```diff
@@ -1141,18 +1141,10 @@ def _image_to_url(b64image: bytes):
-
-        return body_html
+        return re.sub(re.compile(r"""
+            # Group 1: element until the base64 data
+            (<[^>]+\b(?:src="|style="[^"]+\burl\((?:&\#34;|"|'|&quot;)?))
+            data:image/[A-Za-z]+;base64,
+            (.*?) # Group 2: base64 image
+            ((?:(?:&\#34;|"|'|&quot;)\))|") # Group 3: closing the property or attribute
```
![image](https://github.com/odoo/odoo/assets/24205914/3aa0264f-ccee-4f58-8f0b-fdcf1db451d0)

Zinston commented on addons/mass_mailing/models/mailing.py:
```diff
@@ -1128,31 +1128,22 @@ def _convert_inline_images_to_urls(self, body_html):
+            ((?:(?:&\#34;|"|'|&quot;)\))|") # Group 3: closing the property or attribute
+        """, re.VERBOSE), lambda x: x[1] + self._image_to_url(x[2].encode()) + x[3], body_html)
+
+    def _image_to_url(self, b64image: bytes):
```
Done.

Zinston commented:
@dmo-odoo I added tests for the `"` and `&#34;` cases. Then I noticed a few things, added tests for them and addressed them:
1. `style='...'`<br>**Problem**: If we have `url("` it's likely we have `style='` but we only handled `style="`.<br>**Solution**: `style="[^"]+` becomes `style=["'][^"']+`.
2. `url=(...)`<br>**Problem**: `url(` can contain the url without wrapping it in quotes at all.<br>**Solution**: In group 3, `(?:&\#34;|"|'|&quot;)` becomes `(?:&\#34;|"|'|&quot;)?`.
3. `style="...url(...); ...url(...);"`<br>**Problem**: `re.sub` doesn't replace recursively so if there are matches within matches it will only replace once. This means we didn't handle cases where the element's style contained _several_ base64 urls.<br>**Solution**: `return re.sub` becomes `(body_html, do_match) = re.subn` in a `while` loop until nothing is matched anymore.
